syntax = "proto3";

package example;

import "buf/validate/validate.proto";

// Example service definition
service ExampleService {
  // Get a user by ID
  rpc GetUser(GetUserRequest) returns (GetUserResponse);
  
  // Create a new user
  rpc CreateUser(CreateUserRequest) returns (CreateUserResponse);
  
  // List users with pagination
  rpc ListUsers(ListUsersRequest) returns (ListUsersResponse);
}

// Request/Response messages
message GetUserRequest {
  int32 user_id = 1 [(buf.validate.field).int32.gt = 0];
}

message GetUserResponse {
  User user = 1;
}

message CreateUserRequest {
  string name = 1 [(buf.validate.field).string = {
    min_len: 1,
    max_len: 100
  }];
  string email = 2 [(buf.validate.field).string.email = true];
}

message CreateUserResponse {
  User user = 1;
}

message ListUsersRequest {
  int32 page = 1 [(buf.validate.field).int32.gte = 1];
  int32 page_size = 2 [(buf.validate.field).int32 = {
    gte: 1,
    lte: 100
  }];
}

message ListUsersResponse {
  repeated User users = 1;
  int32 total_count = 2;
}

// Data models
message User {
  int32 id = 1 [(buf.validate.field).int32.gt = 0];
  string name = 2 [(buf.validate.field).string = {
    min_len: 1,
    max_len: 100
  }];
  string surname = 3 [(buf.validate.field).string.max_len = 100];
  string email = 4 [(buf.validate.field).string.email = true];
  int64 created_at = 5 [(buf.validate.field).int64.gte = 0];
  int64 updated_at = 6 [(buf.validate.field).int64.gte = 0];
}